clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java

```diff
@@ -259,11 +250,71 @@ private PollResult processAcknowledgements(long currentTimeMs) {
         return pollResult;
     }
 
+    private Optional<UnsentRequest> maybeBuildRequest(AcknowledgeRequestState acknowledgeRequestState,
+            long currentTimeMs, boolean onCommitAsync) {
+        if (acknowledgeRequestState == null || (!closing && acknowledgeRequestState.acknowledgementsToSend.isEmpty()
+                && acknowledgeRequestState.incompleteAcknowledgements.isEmpty())) {
+            if (onCommitAsync) {
+                isAsyncDone = true;
+            }
+            return Optional.empty();
+        } else if (!acknowledgeRequestState.maybeExpire()) {
+            if (nodesWithPendingRequests.contains(acknowledgeRequestState.nodeId)) {
+                log.debug("Skipping acknowledge request because previous request to {} has not been processed",
+                        acknowledgeRequestState.nodeId);
+            } else {
+                if (acknowledgeRequestState.canSendRequest(currentTimeMs)) {
+                    acknowledgeRequestState.onSendAttempt(currentTimeMs);
+                    if (onCommitAsync) {
+                        isAsyncDone = true;
+                    }
+                    return Optional.of(acknowledgeRequestState.buildRequest(currentTimeMs));
+                }
+            }
+        } else {
+            // Fill in TimeoutException
+            for (TopicIdPartition tip : acknowledgeRequestState.incompleteAcknowledgements.keySet()) {
+                metricsManager
+                        .recordFailedAcknowledgements(acknowledgeRequestState.getIncompleteAcknowledgementsCount(tip));
+                acknowledgeRequestState.handleAcknowledgeTimedOut(tip);
+            }
+            acknowledgeRequestState.incompleteAcknowledgements.clear();
+            isAsyncDone = true;
+        }
+        return Optional.empty();
+    }
+
+    private boolean areAnyAcknowledgementsLeft() {
+        boolean areAnyAcksLeft = false;
+        Iterator<Map.Entry<Integer, Pair<AcknowledgeRequestState>>> iterator = acknowledgeRequestStates.entrySet()
+                .iterator();
+        while (iterator.hasNext()) {
+            Map.Entry<Integer, Pair<AcknowledgeRequestState>> acknowledgeRequestStatePair = iterator.next();
+            if (isRequestStateEmpty(acknowledgeRequestStatePair.getValue().getFirst())
+                    && isRequestStateEmpty(acknowledgeRequestStatePair.getValue().getSecond())) {
+                areAnyAcksLeft = true;
+            } else if (!closing) {
+                acknowledgeRequestStates.remove(acknowledgeRequestStatePair.getKey());
+            }
+        }
+        if (!acknowledgeRequestStates.isEmpty())
+            areAnyAcksLeft = true;
+        return areAnyAcksLeft;
+    }
+
+    private boolean isRequestStateEmpty(AcknowledgeRequestState acknowledgeRequestState) {
```

Comment:
This could be clearer. I think `ars == null && ars.acksToSend.isEmpty && ars.ia.isEmpty() && ars.ifa.isEmpty()`.

Reply:
You could even make a method `boolean AcknowledgeRequestState.isEmpty()` so that this code was not intimately aware of the three maps within.

Reply:
Yes makes sense, I have added in a method for the same.